flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java

```diff
@@ -0,0 +1,335 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.is;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertThat;
+
+/**
+ * Tests for {@link DefaultScheduler}.
+ */
+public class DefaultSchedulerTest extends TestLogger {
+
+	private static final int TIMEOUT_MS = 1000;
+
+	private static final JobID TEST_JOB_ID = new JobID();
+
+	private ManuallyTriggeredScheduledExecutor taskRestartExecutor = new ManuallyTriggeredScheduledExecutor();
+
+	private ExecutorService executor;
+
+	private ScheduledExecutorService scheduledExecutorService;
+
+	private Configuration configuration;
+
+	private SubmissionTrackingTaskManagerGateway testTaskManagerGateway;
+
+	private TestRestartBackoffTimeStrategy testRestartBackoffTimeStrategy;
+
+	private FailingExecutionVertexOperationsDecorator testExecutionVertexOperations;
+
+	private SimpleSlotProvider slotProvider;
+
+	private ExecutionVertexVersioner executionVertexVersioner;
+
+	@Before
+	public void setUp() throws Exception {
+		executor = Executors.newSingleThreadExecutor();
+		scheduledExecutorService = new DirectScheduledExecutorService();
+
+		configuration = new Configuration();
+		configuration.setString(JobManagerOptions.EXECUTION_FAILOVER_STRATEGY, FailoverStrategyLoader.NO_OP_FAILOVER_STRATEGY);
+		testTaskManagerGateway = new SubmissionTrackingTaskManagerGateway();
+
+		testRestartBackoffTimeStrategy = new TestRestartBackoffTimeStrategy(true, 0);
+
+		testExecutionVertexOperations = new FailingExecutionVertexOperationsDecorator(new DefaultExecutionVertexOperations());
+
+		slotProvider = new SimpleSlotProvider(TEST_JOB_ID, 12, testTaskManagerGateway);
+
+		executionVertexVersioner = new ExecutionVertexVersioner();
+	}
+
+	@After
+	public void tearDown() throws Exception {
+		if (scheduledExecutorService != null) {
+			scheduledExecutorService.shutdownNow();
+			scheduledExecutorService.awaitTermination(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+		}
+
+		if (executor != null) {
+			executor.shutdownNow();
+			executor.awaitTermination(TIMEOUT_MS, TimeUnit.MILLISECONDS);
+		}
+	}
+
+	@Test
+	public void startScheduling() {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+		final JobVertex onlyJobVertex = getOnlyJobVertex(jobGraph);
+
+		createSchedulerAndStartScheduling(jobGraph);
+
+		final List<ExecutionVertexID> deployedExecutionVertices = testTaskManagerGateway.getDeployedExecutionVertices(1, TIMEOUT_MS);
+
+		final ExecutionVertexID executionVertexId = new ExecutionVertexID(onlyJobVertex.getID(), 0);
+		assertThat(deployedExecutionVertices, contains(executionVertexId));
+	}
+
+	@Test
+	public void restartAfterDeploymentFails() {
+		final JobGraph jobGraph = singleNonParallelJobVertexJobGraph();
+		final JobVertex onlyJobVertex = getOnlyJobVertex(jobGraph);
+
+		testExecutionVertexOperations.enableFailDeploy();
```

Comment:
Is this call safe with respect to visibility?

Reply:
I would say it's safe because this call _happens-before_ all scheduler actions:

https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.4.5

Reply:
because there is no concurrency, right?